src/blog/understanding-ecmascript-part-2.md

```diff
@@ -0,0 +1,305 @@
+>
+> 1. Assert: `IsPropertyKey(P)` is `true`.
+> 2. Let `desc` be `? O.[[GetOwnProperty]](P)`.
+> 3. If `desc` is `undefined`, then
+> a. Let `parent` be `? O.[[GetPrototypeOf]]()`.
+> b. If `parent` is `null`, return `undefined`.
+> c. Return `? parent.[[Get]](P, Receiver)`.
+> 4. If `IsDataDescriptor(desc)` is `true`, return `desc.[[Value]]`.
+> 5. Assert: `IsAccessorDescriptor(desc)` is `true`.
+> 6. Let `getter` be `desc.[[Get]]`.
+> 7. If `getter` is `undefined`, return `undefined`.
+> 8. Return `? Call(getter, Receiver)`.
+
+The prototype chain walk is inside step 3: if we don't find the property as an own property, we recurse into the prototype's `[[Get]]` method.
+
+## Where's the Receiver coming from?
+
+`OrdinaryGet` also passes the original `Receiver` throughout the recursion, unchanged. The `Receiver` is only used if the property is an accessor property, in which case it's passed as the **this value** when calling the getter function.
+
+Let's find out where the `Receiver` is coming from!
+
+Searching for places where `[[Get]]` is called we find an abstract operation `GetValue` which operates on References. Reference is a specification type, consisting of a base value (in this case, an Object), the reference name (in this case, a String), and a strict mode flag (bool).
+
+### Side track: References
+
+Side track: A Reference is not a Record, even though it sounds like it could be &mdash; it contains three fixed named values. Here the spec takes a different approach, and defines References as a higher-level data type. This is because of historical reasons.
+
+### Back to `GetValue`
+
+Let's look at how `GetValue` is defined:
+
+> [`GetValue ( V )`](https://tc39.es/ecma262/#sec-getvalue)
+>
+> 1. `ReturnIfAbrupt(V)`.
+> 2. If `Type(V)` is not `Reference`, return `V`.
+> 3. Let `base` be `GetBase(V)`.
+> 4. If `IsUnresolvableReference(V)` is `true`, throw a `ReferenceError` exception.
+> 5. If `IsPropertyReference(V)` is `true`, then
+> a. If `HasPrimitiveBase(V)` is `true`, then
+> i. Assert: In this case, `base` will never be `undefined` or `null`.
+> ii. Set `base` to `! ToObject(base)`.
+> b. Return `? base.[[Get]](GetReferencedName(V), GetThisValue(V))`.
+> 6. Else,
+> a. Assert: `base` is an Environment Record.
+> b. Return `? base.GetBindingValue(GetReferencedName(V), IsStrictReference(V))`
+
+For property references such as `o.foo`, we take the branch 5 b. The `Receiver` we pass is `GetThisValue(V)`. In this case, it's just the base of the Reference:
+
+> [`GetThisValue( V )`](https://tc39.es/ecma262/#sec-getthisvalue)
+>
+> 1. Assert: `IsPropertyReference(V)` is `true`.
+> 2. If `IsSuperReference(V)` is `true`, then
+> a. Return the value of the `thisValue` component of the reference `V`.
+> 3. Return `GetBase(V)`.
+
+Interesting! The `this` inside the getter refers to the original object where we tried to get the property from, not the one where we found the property during the prototype chain walk. Let's try it out!
```

Comment:
I don't get it. Do we take `2.a` here? What is `IsSuperReference`?

Reply:
Clarified: For `o2.foo`, we don't take the branch in step 2, since it's not a super reference (such as `super.foo`), but we take step 3 and return the base value of the reference which is `o2`.